File: src/main/java/core/file/xml/XMLArenaParser.java
package core.file.xml;

import core.util.HODateTime;
import core.util.HOLogger;
import hattrickdata.*;
import org.apache.commons.lang3.tuple.Pair;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

public class XMLArenaParser {

    private static final String ELEMENT_NAME_FILE_NAME = "FileName";
    private static final String ELEMENT_NAME_VERSION = "Version";
    private static final String ELEMENT_NAME_USER_ID = "UserID";
    private static final String ELEMENT_NAME_FETCHED_DATE = "FetchedDate";
    private static final String ELEMENT_NAME_ARENA = "Arena";
    private static final String ELEMENT_NAME_ARENA_ID = "ArenaID";
    private static final String ELEMENT_NAME_ARENA_NAME = "ArenaName";
    private static final String ELEMENT_NAME_TEAM = "Team";
    private static final String ELEMENT_NAME_TEAM_ID = "TeamID";
    private static final String ELEMENT_NAME_TEAM_NAME = "TeamName";
    private static final String ELEMENT_NAME_LEAGUE = "League";
    private static final String ELEMENT_NAME_LEAGUE_ID = "LeagueID";
    private static final String ELEMENT_NAME_LEAGUE_NAME = "LeagueName";
    private static final String ELEMENT_NAME_REGION = "Region";
    private static final String ELEMENT_NAME_REGION_ID = "RegionID";
    private static final String ELEMENT_NAME_REGION_NAME = "RegionName";
    private static final String ELEMENT_NAME_CURRENT_CAPACITY = "CurrentCapacity";
    private static final String ELEMENT_NAME_CURRENT_CAPACITY_REBUILT_DATE = "RebuiltDate";

    private static final String ELEMENT_NAME_EXPANDED_CAPACITY = "ExpandedCapacity";
    private static final String ELEMENT_NAME_EXPANDED_CAPACITY_EXPANSION_DATE = "ExpansionDate";

    private static final String ATTRIBUTE_NAME_CAPACITY_AVAILABLE = "Available";
    private static final String ELEMENT_NAME_CAPACITY_TERRACES = "Terraces";
    private static final String ELEMENT_NAME_CAPACITY_BASIC = "Basic";
    private static final String ELEMENT_NAME_CAPACITY_ROOF = "Roof";
    private static final String ELEMENT_NAME_CAPACITY_VIP = "VIP";
    private static final String ELEMENT_NAME_CAPACITY_TOTAL = "Total";

    private XMLArenaParser() {
        throw new UnsupportedOperationException("This is a utility class and cannot be instantiated");
    }

    public static Pair<HattrickDataInfo, Arena> parseArenaFromString(String str) {
        return parseDetails(XMLManager.parseString(str));
    }

    private static Pair<HattrickDataInfo, Arena> parseDetails(Document doc) {
        if (doc == null) {
            return null;
        }

        try {
            var hattrickDataInfoBuilder = HattrickDataInfo.builder();

            Element root = doc.getDocumentElement();

            // FileName
            Element element = (Element) root.getElementsByTagName(ELEMENT_NAME_FILE_NAME).item(0);
            hattrickDataInfoBuilder.fileName(XMLManager.getFirstChildNodeValue(element));
            // Version
            element = (Element) root.getElementsByTagName(ELEMENT_NAME_VERSION).item(0);
            hattrickDataInfoBuilder.version(XMLManager.getFirstChildNodeValue(element));
            // UserId
            element = (Element) root.getElementsByTagName(ELEMENT_NAME_USER_ID).item(0);
            hattrickDataInfoBuilder.userId(Integer.parseInt(XMLManager.getFirstChildNodeValue(element)));
            // FetchedDate
            element = (Element) root.getElementsByTagName(ELEMENT_NAME_FETCHED_DATE).item(0);
            hattrickDataInfoBuilder.fetchedDate(HODateTime.fromHT(XMLManager.getFirstChildNodeValue(element)));

            final var hattrickDataInfo = hattrickDataInfoBuilder.build();

            // Root wechseln
            root = (Element) root.getElementsByTagName(ELEMENT_NAME_ARENA).item(0);
            element = (Element) root.getElementsByTagName(ELEMENT_NAME_ARENA_ID).item(0);

            var arenaBuilder = Arena.builder();
            arenaBuilder.id(Integer.parseInt(XMLManager.getFirstChildNodeValue(element)));
            element = (Element) root.getElementsByTagName(ELEMENT_NAME_ARENA_NAME).item(0);
            arenaBuilder.name(XMLManager.getFirstChildNodeValue(element));

            // Team
            Element tmpRoot = (Element) root.getElementsByTagName(ELEMENT_NAME_TEAM).item(0);
            var teamBuilder = Team.builder();
            element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_TEAM_ID).item(0);
            teamBuilder.id(Integer.parseInt(XMLManager.getFirstChildNodeValue(element)));
            element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_TEAM_NAME).item(0);
            teamBuilder.name(XMLManager.getFirstChildNodeValue(element));
            arenaBuilder.team(teamBuilder.build());

            // League
            var leagueBuilder = League.builder();
            tmpRoot = (Element) root.getElementsByTagName(ELEMENT_NAME_LEAGUE).item(0);
            element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_LEAGUE_ID).item(0);
            leagueBuilder.id(Integer.parseInt(XMLManager.getFirstChildNodeValue(element)));
            element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_LEAGUE_NAME).item(0);
            leagueBuilder.name(XMLManager.getFirstChildNodeValue(element));
            arenaBuilder.league(leagueBuilder.build());

            // Region
            var regionBuilder = Region.builder();
            tmpRoot = (Element) root.getElementsByTagName(ELEMENT_NAME_REGION).item(0);
            element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_REGION_ID).item(0);
            regionBuilder.id(Integer.parseInt(XMLManager.getFirstChildNodeValue(element)));
            element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_REGION_NAME).item(0);
            regionBuilder.name(XMLManager.getFirstChildNodeValue(element));
            arenaBuilder.region(regionBuilder.build());

            // Current Capacity
            tmpRoot = (Element) root.getElementsByTagName(ELEMENT_NAME_CURRENT_CAPACITY).item(0);
            element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_CURRENT_CAPACITY_REBUILT_DATE).item(0);

            var currentCapacityBuilder = CurrentCapacity.builder();
            final boolean rebuiltDateAvailable = getXmlAttributeAsBoolean(element, ATTRIBUTE_NAME_CAPACITY_AVAILABLE);
            if (rebuiltDateAvailable) {
                currentCapacityBuilder.rebuildDate(HODateTime.fromHT(XMLManager.getFirstChildNodeValue(element)));
            }

            element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_CAPACITY_TERRACES).item(0);
            currentCapacityBuilder.terraces(Integer.parseInt(XMLManager.getFirstChildNodeValue(element)));
            element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_CAPACITY_BASIC).item(0);
            currentCapacityBuilder.basic(Integer.parseInt(XMLManager.getFirstChildNodeValue(element)));
            element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_CAPACITY_ROOF).item(0);
            currentCapacityBuilder.roof(Integer.parseInt(XMLManager.getFirstChildNodeValue(element)));
            element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_CAPACITY_VIP).item(0);
            currentCapacityBuilder.vip(Integer.parseInt(XMLManager.getFirstChildNodeValue(element)));
            element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_CAPACITY_TOTAL).item(0);
            currentCapacityBuilder.total(Integer.parseInt(XMLManager.getFirstChildNodeValue(element)));

            arenaBuilder.currentCapacity(currentCapacityBuilder.build());

            tmpRoot = (Element) root.getElementsByTagName(ELEMENT_NAME_EXPANDED_CAPACITY).item(0);

            final boolean expandedCapacityAvailable = getXmlAttributeAsBoolean(tmpRoot, ATTRIBUTE_NAME_CAPACITY_AVAILABLE);
            if (expandedCapacityAvailable) {
                var expandedCapacityBuilder = ExpandedCapacity.builder();
                element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_EXPANDED_CAPACITY_EXPANSION_DATE).item(0);
                expandedCapacityBuilder.expansionDate(HODateTime.fromHT(XMLManager.getFirstChildNodeValue(element)));
                element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_CAPACITY_TERRACES).item(0);
                expandedCapacityBuilder.terraces(Integer.parseInt(XMLManager.getFirstChildNodeValue(element)));
                element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_CAPACITY_BASIC).item(0);
                expandedCapacityBuilder.basic(Integer.parseInt(XMLManager.getFirstChildNodeValue(element)));
                element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_CAPACITY_ROOF).item(0);
                expandedCapacityBuilder.roof(Integer.parseInt(XMLManager.getFirstChildNodeValue(element)));
                element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_CAPACITY_VIP).item(0);
                expandedCapacityBuilder.vip(Integer.parseInt(XMLManager.getFirstChildNodeValue(element)));
                element = (Element) tmpRoot.getElementsByTagName(ELEMENT_NAME_CAPACITY_TOTAL).item(0);
                expandedCapacityBuilder.total(Integer.parseInt(XMLManager.getFirstChildNodeValue(element)));

                arenaBuilder.expandedCapacity(expandedCapacityBuilder.build());
            }

            return Pair.of(hattrickDataInfo, arenaBuilder.build());
        } catch (Exception e) {
            HOLogger.instance().log(XMLArenaParser.class, e);
        }

        return null;
    }

    @SuppressWarnings(value = "SameParameterValue")
    private static boolean getXmlAttributeAsBoolean(Element element, String attributeName) {
        return Boolean.parseBoolean(XMLManager.getAttributeValue(element, attributeName).trim());
    }
}

File: src/main/java/core/file/xml/XMLArenaParser.kt
/*
 * XMLArenaParser.java
 *
 * Created on 5. Juni 2004, 15:40
 */
package core.file.xml

import core.util.HOLogger

import org.w3c.dom.Document
import org.w3c.dom.Element

/**
 * 
 * @author thetom
 */
object XMLArenaParser {

	fun parseArenaFromString(str: String): Map<String, String> {
		return parseDetails(XMLManager.parseString(str))
	}

	private fun parseDetails(doc: Document?): Map<String, String> {
		val map = SafeInsertMap()

		if (doc == null) {
			return map
		}

		try {
			var root = doc.documentElement
			var ele = root.getElementsByTagName("FetchedDate").item(0) as Element
			map.insert("FetchedDate", XMLManager.getFirstChildNodeValue(ele))

			// Change of root
			root = root.getElementsByTagName("Arena").item(0) as Element
			ele = root.getElementsByTagName("ArenaID").item(0) as Element
			map.insert("ArenaID", XMLManager.getFirstChildNodeValue(ele))
			ele = root.getElementsByTagName("ArenaName").item(0) as Element
			map.insert("ArenaName", (XMLManager.getFirstChildNodeValue(ele)))

			var tmpRoot = root.getElementsByTagName("Team").item(0) as Element
			ele = tmpRoot.getElementsByTagName("TeamID").item(0) as Element
			map.insert("TeamID", (XMLManager.getFirstChildNodeValue(ele)))
			ele = tmpRoot.getElementsByTagName("TeamName").item(0) as Element
			map.insert("TeamName", (XMLManager.getFirstChildNodeValue(ele)))

			tmpRoot = root.getElementsByTagName("League").item(0) as Element
			ele = tmpRoot.getElementsByTagName("LeagueID").item(0) as Element
			map.insert("LeagueID", (XMLManager.getFirstChildNodeValue(ele)))
			ele = tmpRoot.getElementsByTagName("LeagueName").item(0) as Element
			map.insert("LeagueName", (XMLManager.getFirstChildNodeValue(ele)))

			tmpRoot = root.getElementsByTagName("Region").item(0) as Element
			ele = tmpRoot.getElementsByTagName("RegionID").item(0) as Element
			map.insert("RegionID", (XMLManager.getFirstChildNodeValue(ele)))
			ele = tmpRoot.getElementsByTagName("RegionName").item(0) as Element
			map.insert("RegionName", (XMLManager.getFirstChildNodeValue(ele)))

			tmpRoot = root.getElementsByTagName("CurrentCapacity").item(0) as Element
			ele = tmpRoot.getElementsByTagName("RebuiltDate").item(0) as Element

			if (XMLManager.getAttributeValue(ele, "Available").trim().equals("true", true)) {
				map.insert("RebuiltDate", (XMLManager.getFirstChildNodeValue(ele)))
			}

			ele = tmpRoot.getElementsByTagName("Terraces").item(0) as Element
			map.insert("Terraces", (XMLManager.getFirstChildNodeValue(ele)))
			ele = tmpRoot.getElementsByTagName("Basic").item(0) as Element
			map.insert("Basic", (XMLManager.getFirstChildNodeValue(ele)))
			ele = tmpRoot.getElementsByTagName("Roof").item(0) as Element
			map.insert("Roof", (XMLManager.getFirstChildNodeValue(ele)))
			ele = tmpRoot.getElementsByTagName("VIP").item(0) as Element
			map.insert("VIP", (XMLManager.getFirstChildNodeValue(ele)))
			ele = tmpRoot.getElementsByTagName("Total").item(0) as Element
			map.insert("Total", (XMLManager.getFirstChildNodeValue(ele)))

			tmpRoot = root.getElementsByTagName("ExpandedCapacity").item(0) as Element

			if (XMLManager.getAttributeValue(ele, "Available").trim().equals("true", true)) {
				map.insert("isExpanding", "1")
				ele = tmpRoot.getElementsByTagName("ExpansionDate").item(0) as Element
				map.insert("ExpansionDate", (XMLManager.getFirstChildNodeValue(ele)))
				ele = tmpRoot.getElementsByTagName("Terraces").item(0) as Element
				map.insert("ExTerraces", (XMLManager.getFirstChildNodeValue(ele)))
				ele = tmpRoot.getElementsByTagName("Basic").item(0) as Element
				map.insert("ExBasic", (XMLManager.getFirstChildNodeValue(ele)))
				ele = tmpRoot.getElementsByTagName("Roof").item(0) as Element
				map.insert("ExRoof", (XMLManager.getFirstChildNodeValue(ele)))
				ele = tmpRoot.getElementsByTagName("VIP").item(0) as Element
				map.insert("ExVIP", (XMLManager.getFirstChildNodeValue(ele)))
				ele = tmpRoot.getElementsByTagName("Total").item(0) as Element
				map.insert("ExTotal", (XMLManager.getFirstChildNodeValue(ele)))
			} else {
				map.insert("isExpanding", "0")
				map.insert("ExpansionDate", "0")
				map.insert("ExTerraces", "0")
				map.insert("ExBasic", "0")
				map.insert("ExRoof", "0")
				map.insert("ExVIP", "0")
				map.insert("ExTotal", "0")
			}
		} catch (e: Exception) {
			HOLogger.instance().log(XMLArenaParser.javaClass, e)
		}

		return map
	}
}
